### ISSUE TITLE
`before_batch_transfer` and `after_batch_transfer` hooks in LightningDataModule

### ISSUE THREAD

**Issue by rohitgr7** [PyTorchLightning/pytorch-lightning#3399]
## 🚀 Feature
Can we have two additional hooks in LightningDataModule?
Something like `before_batch_transfer_to_device`, `after_batch_transfer_to_device`, although these can be renamed to something else.

### Motivation

- `before_batch_transfer_to_device`: This can help apply transformations or augmentations to a batch before it is transferred to the device.
- `after_batch_transfer_to_device`: This can help apply transformations or augmentations to a batch after it is transferred to the device. For eg we can perform data augmentation in the GPU for a batch using Kornia.

### Alternatives

As of now, it can be done by overriding the `transfer_batch_to_device` hook but a separate hook makes more sense here I think.

If approved I can send a PR :)

**Comment by edenlightning** [comment]
@williamFalcon @nateraw

**Comment by williamFalcon** [comment]
love it.

**Comment by rohitgr7** [comment]
Should I wait once you are done with refactors? @williamFalcon

**Comment by edenlightning** [comment]
@rohitgr7  I think you can go ahead, most of the reactors are done!

**Comment by rohitgr7** [comment]
I need some suggestions with the API. I think wrapping up these 3 functions `on_before_batch_transfer`, `transfer_batch_to_device` and `on_after_batch_transfer` into a single one something like `prepare_batch` would be better here.
```python
def prepare_batch(self, batch, device, *possible_args, **possible_kwargs):
    batch = self.on_before_batch_transfer(batch, *possible_args, **possible_kwargs)
    batch = self.transfer_batch_to_device(batch, device)
    batch = self.on_after_batch_transfer(batch, *possible_args, **possible_kwargs)
```
since `transfer_batch_to_device` is being called at multiple places already and adding these 2 additional calls everywhere isn't a good way. 

Also move `transfer_batch_to_device` into a separate class `DataDeviceHooks` along with 2 new functions and use it as a parent class only for `LightningDataModule`??

What do you think @awaelchli @nateraw?
I am open to suggestions.

**Comment by wolterlw** [comment]
+1 for implementing a broader `prepare_batch(self, batch, device)` function 
Also consider that right now `transfer_batch_to_device` isn't called when running on CPU.
This causes inconsistencies when debugging on local machines without a GPU.

**Comment by williamFalcon** [comment]
@tchaton

**Comment by tchaton** [comment]
Dear All,

Sounds like a great idea to add a prepare_batch (I would prefer a more explicit name such as prepare_batch_after_collate_fn) hooks made of sub-hooks. I think it should be easier with graphs and sparse pointclouds.

Best,
T.C

**Comment by williamFalcon** [comment]
cc @tchaton

**Comment by leifdenby** [comment]
Thanks for adding these hooks! I would like to use these to apply transforms across my entire batch when training. How would I find out if a batch is part of the training, validation or test set? I want to make sure I'm only applying augmentations during training. I had a look to see I could get that from a state variable within the trainer, but I don't think I can access that within methods called by these hooks. Thank you again, it's really a joy porting my code to pytorch-lightning
